src/cargo/util/progress.rs

```diff
@@ -74,6 +74,94 @@ struct Format {
     style: ProgressStyle,
     max_width: usize,
     max_print: usize,
+    taskbar: TaskbarProgress,
+}
+
+/// Taskbar progressbar
+///
+/// Outputs ANSI codes according to the `Operating system commands`.
+/// Currently only works in the Windows Terminal and ConEmu.
+struct TaskbarProgress {
+    show: bool,
+}
+
+/// A taskbar progress value printable as ANSI OSC escape code
+enum TaskbarValue {
+    /// Do not output anything
+    None,
+    /// Remove progress
+    Remove,
+    /// Progress value 0-100
+    Value(f64),
+    /// Indeterminate state (no bar, just animation)
+    Indeterminate,
+}
+
+enum ProgressOutput {
+    /// Print progress without a message
+    PrintNow,
+    /// Progress, message and taskbar progress
+    TextAndTaskbar(String, TaskbarValue),
+    /// Only taskbar progress, no message and no text progress
+    Taskbar(TaskbarValue),
+}
+
+impl TaskbarProgress {
+    /// Creates a new `TaskbarProgress` from a cargo's config system.
+    ///
+    /// * `config == None` enables taskbar progress reporting on Windows
+    /// by default and disables on other platforms
+    fn from_config(config: Option<bool>) -> Self {
+        let show = match config {
+            Some(v) => v,
+            None => cfg!(target_os = "windows"),
+        };
+
+        TaskbarProgress { show }
+    }
+
+    pub fn remove(&self) -> TaskbarValue {
+        if self.show {
+            TaskbarValue::Remove
+        } else {
+            TaskbarValue::None
+        }
+    }
+
+    pub fn value(&self, percent: f64) -> TaskbarValue {
+        if self.show {
+            TaskbarValue::Value(percent)
+        } else {
+            TaskbarValue::None
+        }
+    }
+
+    pub fn indeterminate(&self) -> TaskbarValue {
+        if self.show {
+            TaskbarValue::Indeterminate
+        } else {
+            TaskbarValue::None
+        }
+    }
+}
+
+impl std::fmt::Display for TaskbarValue {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // From https://conemu.github.io/en/AnsiEscapeCodes.html#ConEmu_specific_OSC
+        // ESC ] 9 ; 4 ; st ; pr ST
+        // When st is 0: remove progress.
+        // When st is 1: set progress value to pr (number, 0-100).
+        // When st is 2: set error state in taskbar, pr is optional.
```

Comment:
I am hesitant to have a ctrl+c handler but I also have had enough bad experience on Windows with colors being leaked from a program and could see myself being annoyed about progress state leaking as well.

Reply:
Here's my prvious work on ctrl+c handler: https://github.com/rust-lang/cargo/commit/de22a7f4f559f129fb909d67ef073687fc73fb94

The approach works, but then we need to set the handler depending on Progress.state. Can you please advice on this? Do you want to have a handler in `src/bin/cargo/main.rs` ?

Reply:
@ehuss I'd be interested in your opinion on this.

Reply:
Since we're now detecting a terminal regardless of the OS, a ctrl+c handler needs to be portable too, not like in my first implementation. If we still want to go this way, a few questions:
1. Can we use external crate like https://crates.io/crates/ctrlc? I assume your policy is to use as few external crates as possible
2. Will it iterfere with https://github.com/rust-lang/cargo/blob/master/crates/cargo-util/src/process_builder.rs#L259-L276? I don't fully understand when this code is used.